add bottlerocket tools to twoliter binary

webern opened the pull request:
*Issue #, if available:*

Closes #14 
Closes #9 
Closes #32
Partly Addresses #43 (We install it here, but should probably still eliminate it in favor of a lib)

*Description of changes:*

Embed tools such as `buildsys`, and scripts such as `rpm2img` into the Twoliter binary. Use them when building Bottlerocket

*Testing*

- [x] Build Bottlerocket on an x86_64 host
- [x] Build Bottlerocket on an Arm host

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

webern commented on .cargo/config.toml:
```diff
@@ -0,0 +1,5 @@
+bindeps = true
+
+[env]
+RUSTC_BOOTSTRAP = "1"
```
Improve this.
See https://github.com/axodotdev/cargo-dist/issues/347#issuecomment-1698141019

webern commented on tools/buildsys/src/builder.rs:
```diff
@@ -121,6 +133,11 @@ impl PackageBuilder {
         }
 
+        // Add the tools directory as a secret since it is not in the context directory.
+        args.extend(tools_secret_args(&PathBuf::from(
```
Right now I'm still passing secrets arguments to the docker build command even though they are not needed because Twoliter is providing an image that has the tools FROM scratch.

I like the `FROM scratch` image better than using secrets, just because it feels cleaner. But I need to choose one method and clear away the detritus from the other method.

webern commented on tools/buildsys/src/builder.rs:
```diff
@@ -650,3 +681,30 @@ where
+//     }
+//     Ok(files)
+// }
+
```
```suggestion
```

webern commented on tools/buildsys/src/gomod.rs:
```diff
@@ -188,8 +189,12 @@ fn docker_go(root_dir: &Path, dg_args: &DockerGoArgs) -> Result<()> {
     let arg_string = args.join(" ");
-    let program = root_dir.join("tools/docker-go");
-    println!("program: {}", program.to_string_lossy());
+    // TODO the getenv function returns an error type that is very inconvenient here.
```
I think I figured this out anyway, clean this up.

webern commented on tools/buildsys/src/gomod.rs:
```diff
@@ -198,7 +203,7 @@ fn docker_go(root_dir: &Path, dg_args: &DockerGoArgs) -> Result<()> {
 
     let stdout = String::from_utf8_lossy(&output.stdout);
-    println!("{}", &stdout);
+    eprintln!("{}", &stdout);
```
Revert this.

webern commented on tools/buildsys/src/builder/error.rs:
```diff
@@ -59,6 +59,9 @@ pub(crate) enum Error {
         source: std::io::Error,
     },
 
+    #[snafu(display("{}", source))]
+    GoModError { source: crate::gomod::error::Error },
+
```
```suggestion
```

webern commented on tools/buildsys/src/gomod/error.rs:
```diff
@@ -35,6 +35,11 @@ pub(crate) enum Error {
         source: std::io::Error,
     },
 
+    /// The error type returned by getenv, located currently in main.rs, cannot be handled properly
+    /// because it would be recursive here.
+    #[snafu(display("{}", message))]
+    MainError { message: String },
+
```
```suggestion
```

webern commented on tools/buildsys/src/gomod/error.rs:
```diff
@@ -55,3 +60,11 @@ pub(crate) enum Error {
 }
 
 pub(super) type Result<T> = std::result::Result<T, Error>;
+
+impl From<crate::error::Error> for Error {
+    fn from(value: crate::error::Error) -> Self {
+        Error::MainError {
+            message: value.to_string(),
+        }
+    }
+}
```
```suggestion
```

webern commented on twoliter/build.rs:
```diff
@@ -10,57 +10,128 @@ Prepare and package embedded tools in a tarball to be included with Twoliter.
+    data_input_dir: PathBuf,
+    /// The directory that we will copy everything to before creating a tarball.
+    prep_dir: PathBuf,
+    /// The path to
```
```suggestion
    /// The path to a file containing the hash of the `Twoliter.dockerfile` image input.
```

webern commented on twoliter/embedded/Dockerfile:
```diff
@@ -9,12 +9,16 @@
 ARG GOARCH
 
-FROM ${SDK} as sdk
+FROM ${TOOLS_IMAGE} as tools
 FROM --platform=linux/${GOARCH} ${TOOLCHAIN}-${ARCH} as toolchain
 
+FROM ${SDK} as sdk
+COPY --chown=1000:1000 --from=tools /tools/ /tools/
```
```suggestion
FROM ${TOOLS_IMAGE} as tools

FROM ${SDK} as sdk
COPY --chown=1000:1000 --from=tools /tools/ /tools/

FROM --platform=linux/${GOARCH} ${TOOLCHAIN}-${ARCH} as toolchain
```

webern commented on twoliter/embedded/Makefile.toml:
```diff
@@ -2,6 +2,10 @@
 skip_core_tasks = true
 
 [env]
+# The directory that twoliter has installed tools like the buildsys binary,
+# rpm2img, etc. There can be no reasonable default for this, it must be
+# specified when running cargo make.
```
Also validate that it is not empty.

```suggestion
# The directory into which Twoliter has installed tools (e.g. buildsys,
# rpm2img, etc.) There can be no reasonable default for this, it must be
# specified when running cargo make.
```

webern commented on twoliter/src/cmd/make.rs:
```diff
@@ -64,6 +68,17 @@ impl Make {
     }
 }
 
+// TODO - remove
```
?

webern commented on twoliter/src/docker/image.rs:
```diff
@@ -32,6 +38,47 @@ impl ImageUri {
         }
     }
+
+    /// Uses `docker image inspect` to discover whether the image exists on the local host.
```
```suggestion
    /// Uses `docker inspect` to discover whether the image exists on the local host.
```

webern commented on twoliter/src/tools/mod.rs:
```diff
@@ -0,0 +1,110 @@
+/// Twoliter project builds, or multiple checkouts of the same Twoliter project.
+fn instance_token(project_dir: impl AsRef<Path>) -> Result<String> {
+    let path = canonicalize(project_dir)?;
+    // Compute a per-checkout prefix for the tag to avoid collisions.
```
```suggestion
```

webern commented:
> https://github.com/bottlerocket-os/twoliter/compare/03e54a7a9a6f157f339e99d65a811bc50b1d09e1..b6c81c329538bc1259413081d604fd2646419f66

Clean some things up.

bcressey commented on tools/buildsys/src/builder.rs:
```diff
@@ -83,6 +85,15 @@ lazy_static! {
 
 static DOCKER_BUILD_MAX_ATTEMPTS: NonZeroU16 = nonzero!(10u16);
 
+/// The list of tools.
+const TOOLS: &[&str] = &[
+    "docker-go",
+    "partyplanner",
+    "rpm2img",
+    "rpm2kmodkit",
+    "rpm2migrations",
+];
```
nit: the buildsys changes might be better as their own commits

bcressey commented on twoliter/embedded/Dockerfile:
```diff
@@ -13,6 +13,18 @@ ARG ARCH
+     --mount=type=secret,id=tools-rpm2migrations,mode=755,target=/tmp/tools/rpm2migrations \
+     cp -a /tmp/tools/ /tools/ \
+     && chown -R 1000:1000 /tools/ \
+     && chmod 755 /tools/
```
nit: are the `chown` and `chmod` necessary?

bcressey commented on twoliter/embedded/Makefile.toml:
```diff
@@ -233,6 +234,17 @@ fi
+# Ensures TWOLITER_TOOLS_DIR is non-zero-length. This variable is unused.
+UNUSED_TWOLITER_TOOLS_DIR_CHECK = { script = [
+'''
+if [ -z "${TWOLITER_TOOLS_DIR}" ];then
+   echo "TWOLITER_TOOLS_DIR must be defined and must be non-zero in length."
+   exit 1
+fi
```
Could we roll this into `tasks.setup` like the architecture check?

bcressey commented on twoliter/embedded/Makefile.toml:
```diff
@@ -464,19 +461,6 @@ rc=0
 export VARIANT="${BUILDSYS_VARIANT}"
 
 # For rust first-party source code
-if ! docker run --rm \
-   -u $(id -u):$(id -g) \
-   -e CARGO_HOME="/tmp/.cargo" \
-   -v "${CARGO_HOME}":/tmp/.cargo \
-   -v "${BUILDSYS_ROOT_DIR}/sources":/tmp/sources \
-   -v "${BUILDSYS_ROOT_DIR}/tools":/tmp/tools \
-   "${BUILDSYS_SDK_IMAGE}" \
-   cargo clippy \
-  --manifest-path /tmp/tools/Cargo.toml \
-  --locked -- -D warnings --no-deps; then
```
The `check-shell` test was previously linting some of the migrated scripts so we should bring that back to this repo at some point.

webern replied:
> The `check-shell` test was previously linting some of the migrated scripts so we should bring that back to this repo at some point.

I have an issue open for this https://github.com/bottlerocket-os/twoliter/issues/42

bcressey commented on twoliter/embedded/Makefile.toml:
```diff
@@ -939,25 +867,12 @@ dependencies = [
 set -e
 
-export PATH="${BUILDSYS_TOOLS_DIR}/bin:${PATH}"
+export PATH="${TWOLITER_TOOLS_DIR}:${PATH}"
```
Are we including `infrasys` also? I'd be OK with just dropping it here, not including it as a tool, and removing the sources later.

webern replied:
> Are we including `infrasys` also? I'd be OK with just dropping it here, not including it as a tool, and removing the sources later.

I moved the sourcecode to here but I am not embedding it in twoliter. I will remove this makefile target.

bcressey commented on twoliter/build.rs:
```diff
@@ -11,56 +11,108 @@ use bytes::BufMut;
 
-    // This is the filepath to the tarball we will create.
-    let tar_path = out_dir.join("tools.tar.gz");
+    paths.copy_file("Dockerfile");
+    paths.copy_file("Makefile.toml");
+    paths.copy_file("docker-go");
+    paths.copy_file("partyplanner");
+    paths.copy_file("rpm2img");
+    paths.copy_file("rpm2kmodkit");
+    paths.copy_file("rpm2migrations");
```
Do we need to emit `rerun-if-changed` for the non-Rust binary artifacts?

webern replied:
> Do we need to emit `rerun-if-changed` for the non-Rust binary artifacts?

I am! `println!("cargo:rerun-if-changed={}", paths.data_input_dir.display());`

cbgbt commented on .github/workflows/rust.yml:
```diff
@@ -18,6 +18,8 @@ on:
   build:
     runs-on: ubuntu-latest
+    env:
+      RUSTC_BOOTSTRAP: "1"
```
I've never seen this before, probably worth a comment here as to why. Is this to avoid using the nightly compiler?

webern committed:
add bottlerocket build tools to twoliter

Adds Rust tools and scripts, originally located in the Bottlerocket
repo, to Twoliter's embedded tools.

webern committed:
remove an unused item from .gitignore

webern committed:
buildsys: integrate with embedded twoliter tools

Buildsys requires these changes to work with the tools embedded in the
Twoliter binary.

webern commented:
> https://github.com/bottlerocket-os/twoliter/compare/6ebb528c1b64159e6242e897d08784e82381b069..8a945cf94116f92a32ab08191bb0410c298794d6

Split commits

webern commented:
> https://github.com/bottlerocket-os/twoliter/compare/8a945cf94116f92a32ab08191bb0410c298794d6..c6b7e0961a849d2fbd6a330457a9efa2794090f9

Fixup bcressey's suggestions

webern committed:
add debug command to twoliter

Adds a command that can be used to dump the tools for further
inspection.

webern commented:
> https://github.com/bottlerocket-os/twoliter/compare/5406dc0f0187d491c24103f23a6c2f29d2cf3376..14a9919c7609da32e8551c03558104d7c09a608d

Substantial rework of how the tools are embedded. As it turns out, and this makes sense, `build-dependencies` are always built for the host machine. We need the `bindeps` to be `dependencies` so they will be built for the `--target`.

I cannot figure out a way to compress these bytes right now. The `include-flate` macro requires the path to be an actual string literal (not a macro like `env!` that evaluates to a string literal). This might be an oversight (not sure how it all works) that is fixable. But also note that the compilation of `include-flate` seemed to take forever. I'm not sure what is going on there.

webern committed:
add version flag to twoliter

Enable printing the current version with twoliter --version

webern committed:
use large github actions runners

cbgbt commented on rust-toolchain.toml:
```diff
@@ -0,0 +1,6 @@
+# This file is locks the build to a specific nightly compiler. There is nothing special about the
```
nit

```suggestion
# This file locks the build to a specific nightly compiler. There is nothing special about the
```

bcressey reviewed (commented):
Trying to reduce `twoliter` size a bit, I tested [combinations of these settings](https://gist.github.com/bcressey/a9e6d256dc4c6e1a53d13ce58d527ac9) in `Cargo.toml`:
```
[profile.release]
strip = "debuginfo"
codegen-units = 1
lto = true
```

At a bare minimum, I'd recommend stripping debuginfo.

Setting codegen units to 1 is the biggest one but costs the most in terms of added time. If we go there then we may as well go full LTO for the extra 18% size reduction. This seems worthwhile if our release builders can handle it, but definitely not for CI.

webern commented:
> https://github.com/bottlerocket-os/twoliter/compare/552f9243c9d7282219fe53e4cd33aab5a821cc6e..9af3a998650d46758e1b34990e39fc23f5f7a7f7

Add release profile with one codegen unit.

webern commented:
> https://github.com/bottlerocket-os/twoliter/compare/9af3a998650d46758e1b34990e39fc23f5f7a7f7..84b795a80985ea12524b7d6a75d37e2ac3ef1eb6

Rebase over the latest tools changes from Bottlerocket.

Includes:
- rpm2img xfs fix
- testsys changes

webern commented:
> At a bare minimum, I'd recommend stripping debuginfo.

Nice. This change took the binary from 380M to 80M!

webern commented:
> https://github.com/bottlerocket-os/twoliter/compare/84b795a80985ea12524b7d6a75d37e2ac3ef1eb6..158e5179fe5e55e39c5e0b505ba598fc8e0aeee7

Fix a typo found by Sean.

webern committed:
stop printing the entire cargo make command

It is not a good idea to print the entire cargo make command when an
error occurs. This is making it harder for people to find the real'
error and could reveal secrets. Best to be quiet about what command was
invoked under the hood.

webern commented:
> https://github.com/bottlerocket-os/twoliter/pull/47/commits/a527e076c885834efd25a22479c714f4cec21c70

Tacked on a commit to stop the unhelpful printing of the entire cargo make invocation which makes it harder for people to find the real error.

webern commented:
> https://github.com/bottlerocket-os/twoliter/compare/a527e076c885834efd25a22479c714f4cec21c70..ace7b30e6400df5a9af943dfe25b4ebd2cf10536

Rebase to pickup changes from Bottlerocket.
